applications/sdp/mspi/src/hrt/hrt.c

```diff
@@ -6,180 +6,108 @@
-
-	nrf_vpr_csr_vio_mode_out_set(&out_mode);
-	nrf_vpr_csr_vio_mode_in_buffered_set(NRF_VPR_CSR_VIO_MODE_IN_CONTINUOUS);
 
-	nrf_vpr_csr_vio_config_get(&config);
-	config.input_sel = false;
-	nrf_vpr_csr_vio_config_set(&config);
-
-	/* Fix position of data if word size < MAX_WORD_SIZE,
-	 * so that leading zeros would not be printed instead of data bits.
-	 */
-	if (xfer_ll_params.word_size < MAX_WORD_SIZE) {
-		for (uint8_t i = 0; i < xfer_ll_params.data_len; i++) {
-			xfer_ll_params.data_to_send[i] =
-				xfer_ll_params.data_to_send[i]
-				<< (MAX_WORD_SIZE - xfer_ll_params.word_size);
-		}
+	if (xfer_ll_params.ce_polarity == MSPI_CE_ACTIVE_LOW) {
+		out = BIT_SET_VALUE(out, xfer_ll_params.ce_vio, VPRCSR_NORDIC_OUT_LOW);
+	} else {
+		out = BIT_SET_VALUE(out, xfer_ll_params.ce_vio, VPRCSR_NORDIC_OUT_HIGH);
 	}
-
-	/* Counter settings */
-	nrf_vpr_csr_vtim_count_mode_set(0, NRF_VPR_CSR_VTIM_COUNT_RELOAD);
-	nrf_vpr_csr_vtim_simple_counter_top_set(0, xfer_ll_params.counter_top);
-
-	/* Set number of shifts before OUTB needs to be updated.
-	 * First shift needs to be increased by 1.
-	 */
-	nrf_vpr_csr_vio_shift_cnt_out_set(xfer_ll_params.word_size);
-	nrf_vpr_csr_vio_shift_cnt_out_buffered_set(xfer_ll_params.word_size - 1);
-
-	/* Enable CS */
-	out = nrf_vpr_csr_vio_out_get();
-	out &= ~PIN_OUT_HIGH_MASK(VIO(NRFE_MSPI_CS0_PIN_NUMBER));
-	out |= xfer_ll_params.ce_enable_state ? PIN_OUT_HIGH_MASK(VIO(NRFE_MSPI_CS0_PIN_NUMBER))
-					      : PIN_OUT_LOW_MASK(VIO(NRFE_MSPI_CS0_PIN_NUMBER));
 	nrf_vpr_csr_vio_out_set(out);
 
-	/* Start counter */
-	nrf_vpr_csr_vtim_simple_counter_set(0, CLK_FIRST_CYCLE_MULTIPLICATOR *
-						       xfer_ll_params.counter_top);
+	for (uint32_t i = 0; i < xfer_ll_params.words; i++) {
+		switch (xfer_ll_params.words - i) {
+		case 1: /* Last transfer */
+			nrf_vpr_csr_vio_shift_cnt_out_buffered_set(xfer_ll_params.last_word_clocks -
+								   1);
+			switch (xfer_ll_params.bit_order) {
+			case HRT_BO_NORMAL:
+				nrf_vpr_csr_vio_out_buffered_set(xfer_ll_params.last_word);
+				break;
+			case HRT_BO_REVERSED_BYTE:
+				nrf_vpr_csr_vio_out_buffered_reversed_byte_set(
+					xfer_ll_params.last_word);
+				break;
+			case HRT_BO_REVERSED_WORD:
+				nrf_vpr_csr_vio_out_buffered_reversed_word_set(
+					xfer_ll_params.last_word);
+				break;
+			}
+			break;
+		case 2: /* Last but one transfer.*/
+			nrf_vpr_csr_vio_shift_cnt_out_buffered_set(
+				xfer_ll_params.penultimate_word_clocks - 1);
+
+			/* Intentional fallthrough. */
+		default:
+			switch (xfer_ll_params.bit_order) {
+			case HRT_BO_NORMAL:
+				nrf_vpr_csr_vio_out_buffered_set(
+					((uint32_t *)xfer_ll_params.data)[i]);
+				break;
+			case HRT_BO_REVERSED_BYTE:
+				nrf_vpr_csr_vio_out_buffered_reversed_byte_set(
+					((uint32_t *)xfer_ll_params.data)[i]);
+				break;
+			case HRT_BO_REVERSED_WORD:
+				nrf_vpr_csr_vio_out_buffered_reversed_word_set(
+					((uint32_t *)xfer_ll_params.data)[i]);
+				break;
+			}
+		}
```

Comment:
This is configured once for device or transfer. So I think this switch-case should be resolved in configuration phase. Not in real time.
Appropriate `nrf_vpr_csr_vio_out_buffered...` functions should be passed through pointers.